### PR TITLE
chore(attribution): check dependency license and add attributions

### LINKED ISSUE #43: dependency license checking
**Is your feature request related to a problem? Please describe.**
We need to check all the licenses of our dependencies to ensure they are compatible with our project

**Describe the solution you'd like**
We need to go through each dependency and check if we are able to use it, and if so are there any conditions to use?
We may have to add an attribution section to our app. Potentially could also attribute in the wiki but unsure

Also should add a check in the PR checklist to add any attribution required

### Checklist:

- [x] I have double checked that the issue does not currently exist. 
- [x] I have added the appropriate labels (e.g. feature, front-end, back-end).
- [x] I have added the **Platform Project** project

### PR DESCRIPTION
**Pull request by wshe874**
# Description

- checking the license of all the third party dependencies used in the project to ensure our MIT license is sufficient for legal requirements.
- adding attributions to all the third party dependencies and images used in the project.

Fixes/resolves #43

## Screenshots
Please include any screenshots or media that illustrates the changes made

## Type of change
Please delete options that are not relevant.

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **Breaking change** (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:
Leave blank if not applicable

I have completed these steps when making this pull request:
- [x] I have assigned my name to the issue
- [x] I have moved the issue to the **In Progress** column
- [x] I have labelled the PR appropriately
- [x] I have assigned myself to the PR

Before opening the PR for review:
- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I have moved the linked issue to the **Review in Progress** column

### CODE REVIEW

**Comment by GlockTian** [comment]
Can you document how you generate these third party license files in Readme?

**Comment by wshe874** [comment]
> Can you document how you generate these third party license files in Readme?

@GlockTian documentations updated in Readme

File: README.md [hunk 1]
```diff
@@ -13,12 +13,27 @@ Core repository for the platform team 4 is developing
...
+1. Run `yarn run start:docker` to simultaneously run the frontend and backend in docker containers.
 
 ## Tests
 
-1. Run `yarn run test:docker` to simultaneously run the frontend and backend tests in docker containers. 
+1. Run `yarn run test:docker` to simultaneously run the frontend and backend tests in docker containers.
 
 For more detail, view `CONTRIBUTING.md`.
+
+# Generate LICENSE-3RD-PARTY.txt
```
Comment (GlockTian) [inline README.md]: This should be a subtitle section? so ##

Comment (wshe874) [reply README.md]: This cannot be a subtitle unfortunately as this is not a part of the "Run with Docker" documentation.

File: packages/frontend/package.json [hunk 2]
```diff
@@ -32,7 +32,8 @@
     "test": "react-scripts test",
     "eject": "react-scripts eject",
     "generate:api-types": "npx openapi-typescript http://localhost:4200/api-json --output ./src/types/api-schema.ts",
-    "generate": "run-p generate:*"
+    "generate": "run-p generate:*",
```
Comment (frasermcc9) [inline packages/frontend/package.json:35]: Place the plain generate script above its subparts

File: package.json [hunk 3]
```diff
@@ -18,6 +18,7 @@
     "start:dev": "lerna run --parallel start:dev",
     "test": "lerna run --parallel test",
     "start:docker": "docker-compose up",
-    "test:docker": "docker-compose -f docker-compose.override.yml up"
+    "test:docker": "docker-compose -f docker-compose.override.yml up",
+    "generate:license-report": "license-report license-report --output=table --config license-report-config.json > LICENSE-3RD-PARTY.txt"
```
Comment (frasermcc9) [inline package.json]: Can this also run the script in the subpackages (using lerna run)

File: packages/frontend/package.json (hunk 2 above)
Comment (j-chad) [inline packages/frontend/package.json:35]: I don't think we want to have a command that does both generate tasks. They are both one-off tasks and not related. By adding this script we are just cluttering the package.json and making it harder for new programmers to grep

Comment (frasermcc9) [reply packages/frontend/package.json:35]: True, it should be done postinstall

**Review by j-chad — COMMENTED** [review]
are these text files enough for attribution? Does it also have to be available on our site?

**Review by frasermcc9 — CHANGES_REQUESTED** [review]
To summarize my suggestions:
- Rename script to just `license-report`
- Add a postintstall script to frontend and backend that runs it
- Modify the root postinstall to run it i.e. `lerna bootstrap && npm run license-report`

Believe we need the postinstall in all packages since the root postinstall wont run if we `npm i` in one of the subpackages.

**Comment by frasermcc9** [comment]
> are these text files enough for attribution? Does it also have to be available on our site?

Strictly speaking no since webpack collects license information and outputs it into the build directory

**Comment by j-chad** [comment]
> Strictly speaking no since webpack collects license information and outputs it into the build directory

Ok well that's probably fine, we have put the legwork in so that if we had to (and were still working on it) in the future we can automatically add to our website

**Comment by wshe874** [comment]
> To summarize my suggestions:
> 
> * Rename script to just `license-report`
> * Add a postintstall script to frontend and backend that runs it
> * Modify the root postinstall to run it i.e. `lerna bootstrap && npm run license-report`
> 
> Believe we need the postinstall in all packages since the root postinstall wont run if we `npm i` in one of the subpack

@frasermcc9 It is nice to have the whole process automated, and I really want it automated too. However, after some experimentation, it seems that the license-report package is extremely slow at producing a report. I have also experimented with a few other tools available such as license-cheker and legally, all of them not only check the third-party dependencies in the package.json but also check the license for the dependencies in the node_module. Therefore, I would not recommend running the script post-install, instead developers should only manually run license-report when necessary.

**Comment by frasermcc9** [comment]
How slow are we talking here

It could be ran as part of CI/CD on merge, and then commits the changes automatically to main.

**Comment by j-chad** [comment]
can you set a maximum depth so it doesn't go into node_modules?

**Comment by j-chad** [comment]
If not, I'm not fussed if we don't do automation

**Comment by j-chad** [comment]
> How slow are we talking here
> 
> It could be ran as part of CI/CD on merge, and then commits the changes automatically to main.

Our CI budget is quite tight as it is. We should only add this if it takes less than 20 seconds

**Comment by wshe874** [comment]
> How slow are we talking here
> 
> It could be ran as part of CI/CD on merge, and then commits the changes automatically to main.

Sometimes it almost seems like not gonna end, had to cancel the process.

**Comment by wshe874** [comment]
> can you set a maximum depth so it doesn't go into node_modules?

For the dependencies that goes through node_modules, we are unable to set a maximum depth.

**Comment by wshe874** [comment]
> > can you set a maximum depth so it doesn't go into node_modules?
> 
> For the dependencies that goes through node_modules, we are unable to set a maximum depth.

Actually I just found another package that I can use to generate a license report.

File: ATTRIBUTIONS.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+The following attributions are for images used in the project:
```
Comment (j-chad) [inline ATTRIBUTIONS.txt]: Can we make this file markdown and format as a markdown table

File: LICENSE-3RD-PARTY.txt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# Dependency Licenses
+## root ()
+-------------------
+
+| Package Name | Version | URL | Description | License |
+| ------------ | ------- | --- | ----------- | ------- |
+| husky | 7.0.4 | https://typicode.github.io/husky | Modern native Git hooks made easy |  MIT |
```
Comment (j-chad) [inline LICENSE-3RD-PARTY.txt]: change file name to `LICENSE-3RD-PARTY.md`

File: packages/backend/LICENSE-3RD-PARTY.txt [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# Dependency Licenses
+## platform-backend (0.1.0)
```
Comment (j-chad) [inline packages/backend/LICENSE-3RD-PARTY.txt]: rename to `LICENSE-3RD-PARTY.md`

File: packages/frontend/LICENSE-3RD-PARTY.txt [hunk 7]
```diff
@@ -0,0 +1,36 @@
+# Dependency Licenses
+## frontend (0.1.0)
+-------------------
+
```
Comment (j-chad) [inline packages/frontend/LICENSE-3RD-PARTY.txt]: rename to `LICENSE-3RD-PARTY.md`

**Review by j-chad — CHANGES_REQUESTED** [review]

**Review by j-chad — APPROVED** [review]
LGTM

**Comment by j-chad** [comment]
oh just a thought: maybe make the root license-report use lerna run to also run the package ones

**Review by j-chad — CHANGES_REQUESTED** [review]
Please add license-report-generator to dev dependencies

**Pull request closed without merging**